vhost_user_backend: Delegate worker thread management to backends

bjzhjing opened the pull request:
While add multiple threads support for vhost_user_net backend, it's found that there is race condition while multiple threads concurrent access and change data structures defined in backend, which lead to performance penalty. To resolve such issue, it's better to let vhost_user_net backend to create and manage the worker threads itself, then those threads will have data structures respectively, we can see throughput improved then. 

Considering that worker threads aim to handle virtqueue data in which each type of backend may have specific requirements, additionally, to keep the design of vhost_user_backend clean, we'd like to move the worker thread management out of the vhost_user_backend lib code, let backends to handle it on their own. 

This PR is opened to implement the new worker thread strategy for vhost_user_backend and the three backend, fs/net/block.

bjzhjing commented:
@rbradford @sboeuf 
There are two failure cases in CI test, which I could not reproduce in my local environment, could you please help see how to recreate ?
failures:
    tests::test_boot_from_vhost_user_blk
    tests::test_vhost_user_blk_self_spawning

rbradford commented on vhost_user_block/src/lib.rs:
```diff
@@ -228,8 +228,8 @@ impl VhostUserBackend for VhostUserBlkBackend {
-    fn set_event_idx(&mut self, enabled: bool) {
-        self.event_idx = enabled;
+    fn set_event_idx(&mut self, _enabled: bool) {
+        self.event_idx = false;
```
This is not a good idea. @slp do you have any idea on what's going on here?

slp replied:
@rbradford That's wrong, it's inhibiting vhost_user_blk EVENT_IDX support.

rbradford replied:
@slp My thoughts exactly! Perhaps you could lend a hand on understanding what's going wrong.

bjzhjing replied:
@slp @rbradford  I comment event_idx just to verify if event_idx is related to the failure of test_boot_from_vhost_user_blk and test_vhost_user_blk_self_spawning. Yes, this is not a fix, but to circumvented the issue temporary.

slp commented:
> While add multiple threads support for vhost_user_net backend, it's found that there is race condition while multiple threads concurrent access and change data structures defined in backend, which lead to performance penalty. To resolve such issue, it's better to let vhost_user_net backend to create and manage the worker threads itself, then those threads will have data structures respectively, we can see throughput improved then.

@bjzhjing Is there a GitHub issue talking about this issue in more detail?

bjzhjing commented:
@sameo @sboeuf @rbradford 
multiple worker thread support is already added for vhost-user-net backend through PR #785, which has gained better performance. As discussed on weekly meeting, considering what is done in PR#785, we decide to move worker thread out of vhost_user_backend and delegate to the backends, so that's why this PR is opened, to let backends create single worker thread to verify if the new design works. 

This PR has passed all CI integration tests, which I think it indicates the above idea to let backends manage worker thread is workable. 

There is one thing need to be mentioned here, that if event_idx is enabled for vhost_user_blk, no matter I set --kernel with vmlinux or hyper-firmware while start guest, the CI integration test cases test_boot_from_vhost_user_blk and test_vhost_user_blk_self_spawning will both fail, once event_idx is disabled (that is what the last patch did), those two cases will pass. I'm digging into this currently, this issue is exposed through this PR, but I think it's not caused by this PR.

I'm talking here because I'd like to confirm with you if we could get back to #785 for a final review. It will be better to have it get merged into next release before the benchmark test start. Let me know what's your plan. Thanks!

bjzhjing commented:
> > While add multiple threads support for vhost_user_net backend, it's found that there is race condition while multiple threads concurrent access and change data structures defined in backend, which lead to performance penalty. To resolve such issue, it's better to let vhost_user_net backend to create and manage the worker threads itself, then those threads will have data structures respectively, we can see throughput improved then.
> 
> @bjzhjing Is there a GitHub issue talking about this issue in more detail?

Not yet, I could open it if you need. I'm currently investigating on this, it will be helpful if you could take a look.

slp commented:
@bjzhjing Actually, I was asking about the original motivation for this patch set. Particularly, about this affirmation:

> While add multiple threads support for vhost_user_net backend, it's found that there is race condition while multiple threads concurrent access and change data structures defined in backend, which lead to performance penalty. 

I'm going to take a look at the various, seemly related PRs to try to understand what the problem is and what this patch series attempts to tackle.

bjzhjing commented:
> @bjzhjing Actually, I was asking about the original motivation for this patch set. Particularly, about this affirmation:
> 
> > While add multiple threads support for vhost_user_net backend, it's found that there is race condition while multiple threads concurrent access and change data structures defined in backend, which lead to performance penalty.
> 
> I'm going to take a look at the various, seemly related PRs to try to understand what the problem is and what this patch series attempts to tackle.

While we have another PR #785 which aims to add multiple thread support for vhost_user_net backend when multiple queue enabled. For net device, considering the cpu affinity, it's better to let one worker thread service one Rx/Tx queue pair, while there is multiple queue pairs configured, there will be multiple worker threads created. Based on the original vhost_user_backend design, those threads will be created by Daemon (Vhost_user_backend), then, access and change the data structures RxVirtio/TxVirtio defined in VhostUserNetBackend. However, there is no performance improved after multiple threads are created. I suppose there is rust inherent lock protection on those data structures. And then, I turn to let vhost_user_net backend create the worker threads on its own, and define structures for each of them. we could see the performance improvement after that. 
With the change in #785, worker threads will be managed or not by vhost_user_backend, it depends on the backends' requirement. @rbradford and @sboeuf have help review the change, and there is a discussion about the new design. Since worker thread is used to handle the data part for each backend which may have specified requirement on that, just like what the net backend encountered, in addition,  to keep the design of vhost_user_backend be uniform and clean, we decide to move the worker thread management out of vhost_user_backend, and delegate to backends completely.  So that's why the PR is opened.

slp commented on vhost_user_backend/src/lib.rs:
```diff
@@ -718,9 +720,15 @@ impl<S: VhostUserBackend> VhostUserSlaveReqHandler for VhostUserHandler<S> {
+                    fd.as_raw_fd(),
+                    epoll::Events::EPOLLIN,
+                    u64::from(index),
+                )
```
If the backend is going to deal with the worker threads, it should also manage the epoll fds itself. I think this should be replaced with a backend call with index and RawFd as arguments.

On the backend, this call can also be a good hint for creating the vring worker, if it doesn't want to create them in advance.

rbradford replied:
I agree with @slp I think the epoll fds should be managed in the backends.

bjzhjing replied:
@rbradford update is done to let backends manage epoll fds.

slp commented on vhost_user_net/src/lib.rs:
```diff
@@ -183,6 +210,11 @@ impl VhostUserBackend for VhostUserNetBackend {
+        if self.vrings.is_some() && self.run().is_err() {
+            return Err(Error::FailedRunHandler.into());
+        }
+
```
Why create the worker thread here? IMO they either should be created in advance, while initializing VhostUserBackend, or even better, when as VHOST_USER_SET_VRING_KICK message is received.

Also, I don't think "run" is a good name for a method that creates a worker thread an returns.

bjzhjing replied:
start_worker_thread is added in vhost_user_backend to notify the backend to start worker thread at the point when memory is set for backend in set_mem_table. If we start the worker thread too early, mem is None, so the epoll handler could not handle event. That's why we start the thread once memory is update. set_mem_table is earlier than VHOST_USER_SET_VRING_KICK received, so I think it meets your expectation.

slp commented on vhost_user_block/src/lib.rs:
```diff
@@ -245,8 +248,10 @@ impl BlockEpollHandler {
                 Ok(request) => {
                     debug!("element is a valid request");
+                    let mut disk_image_locked = self.disk_image.lock().unwrap();
+                    let mut disk_image = disk_image_locked.deref_mut();
```
This would inhibit any significant amount of parallelism in vhost_user_blk. It's fine if this is just an initial implementation to be improved upon in future changes, but if that's the case, I think there should be a TODO comment there.

bjzhjing replied:
Sorry, I just push a new version but missing your comment here, I will add it later.

slp commented on vhost_user_backend/src/lib.rs:
```diff
@@ -622,5 +632,11 @@ impl<S: VhostUserBackend> VhostUserSlaveReqHandler for VhostUserHandler<S> {
+                error!("Error in vring worker: {:?}", e);
+            }
+        }
+    }
```
If backends are in charge of managing the worker threads, why should we join them in VhostUserHandler's Drop? Also, this only deals with a single worker thread.

bjzhjing replied:
Right, remove this part.

slp commented on src/bin/vhost_user_fs.rs:
```diff
@@ -404,6 +397,24 @@ fn main() {
+            break;
+        }
+    }
+
```
Oh, gosh, please no. Worker threads must support updating their properties, when necessary. This is just cheating in an incredible ugly way.

bjzhjing replied:
Yes, I understand it. It's not a good way to handle the thing. I just take a look in vm-virtio/src/vhost_user/fs, and I move the set_slave_req_fd ahead of other vhost_user protocol handling, this could ensure both vu_req and mem are not none when we start worker thread for fs backend. So never need to loop here, just remove this part.

slp commented on vhost_user_block/src/lib.rs:
```diff
@@ -522,6 +517,16 @@ pub fn start_block_backend(backend_command: &str) {
+            break;
+        }
+    }
+
```
This is a terribly ugly hack. If worker threads can't be created in advance for some reason, they must created after receiving a VHOST_USER_SET_VRING_KICK.

bjzhjing replied:
> So I guess this is about avoiding the backend RwLock write lock at VringEpollHandler. That's fine and it's something that we needed to tackle sooner than later.
> 
> I've added some suggestions to the code. There are also a number of commits with a summary but empty description. I think that should be fixed too.

Thanks for helping review this and providing good suggestions, I have addressed them (except one comment to be added). PTAL. But, I see the event_idx issue is still there, do you have any clue? Why this happens?

slp commented on vhost_user_net/src/lib.rs:
```diff
@@ -153,6 +133,53 @@ impl VhostUserNetBackend {
+                rxs,
+                txs,
+                rx_tap_listenings,
+                epoll_fd: self.epoll_fd,
```
Each handler needs its own epoll_fd. You also need a way to map each worker's epoll_fd with one or more vring index, to register the RawFd's when receiving a VHOST_USER_SET_VRING_KICK.

bjzhjing replied:
Since this PR only handle one worker thread, we could extend later by adding vector<epoll_fd>, and map them with vrings in set_vring_epoll_fd based on backends' requirement.

slp commented:
So I guess this is about avoiding the backend RwLock write lock at VringEpollHandler. That's fine and it's something that we needed to tackle sooner than later.

I've added some suggestions to the code. There are also a number of commits with a summary but empty description. I think that should be fixed too.

bjzhjing commented:
> So I guess this is about avoiding the backend RwLock write lock at VringEpollHandler. That's fine and it's something that we needed to tackle sooner than later.
> 
> I've added some suggestions to the code. There are also a number of commits with a summary but empty description. I think that should be fixed too.

Thanks for helping review this and providing good suggestions, I have addressed them (except one comment to be added). PTAL. But, I see the event_idx issue is still there, do you have any clue? Why this happens?

bjzhjing commented:
@rbradford @slp I recreate the failed test case that test_boot_from_vhost_user_block, and find the root cause, I have opened a [PR 34](https://github.com/cloud-hypervisor/rust-hypervisor-firmware/pull/34) to add a fix in [rust-hypervisor-firmware](https://github.com/cloud-hypervisor/rust-hypervisor-firmware/), PTAL. Once fix accepted, I will remove the event_idx disable patch here.

slp commented:
The actual problem is that when `set_event_idx` is called on the `VhostUserBlkBackend`, the only value getting updated is the `event_idx` in the backend itself, but not in the worker thread you spawned with `start_worker_thread`.

As I said before, you can't assume worker properties will be valid for all the lifetime of the backend, so you need a way to update them. This can be done stopping/creating them, calling a method (with proper locking) or event with an event delivered through evenfd or a channel.

bjzhjing commented:
> The actual problem is that when `set_event_idx` is called on the `VhostUserBlkBackend`, the only value getting updated is the `event_idx` in the backend itself, but not in the worker thread you spawned with `start_worker_thread`.
> 
> As I said before, you can't assume worker properties will be valid for all the lifetime of the backend, so you need a way to update them. This can be done stopping/creating them, calling a method (with proper locking) or event with an event delivered through evenfd or a channel.

Yes, I agree! Look at the affect of memory update caused by hotplug, there should be event notification between backend and thread.

bjzhjing commented:
@slp I've update with your good suggestion to notify worker properties change through event. That's really helpful!

bjzhjing commented:
@sboeuf @rbradford @sameo This PR is ready for review, PTAL! Thanks!

rbradford commented:
Have you tried porting network MQ support (#785) to the framework this PR introduces? I think that would be the test of the design.

bjzhjing commented:
Close it for #1012 drive the same thing.
